src/psyclone/tests/psyir/backend/sympy_writer_test.py

```diff
@@ -136,6 +138,74 @@ def test_sym_writer_functions(fortran_reader, expressions):
                 end program test_prog '''.format(expressions[0])
 
     psyir = fortran_reader.psyir_from_source(source)
-    lit = psyir.children[0].children[0].rhs
+    function = psyir.children[0].children[0].rhs
     sympy_writer = SymPyWriter()
-    assert sympy_writer(lit) == expressions[1]
+    assert sympy_writer(function) == expressions[1]
+
+
+@pytest.mark.parametrize("expressions", [("a%x", "a%a_x"),
+                                         ("b(i)%x", "b(i)%b_x"),
+                                         ("a%x(i)", "a%a_x(i)"),
+                                         ("b(j)%x(i)", "b(j)%b_x(i)"),
+                                         ("b(i)%c(b_c)", "b(i)%b_c_1(b_c)"),
+                                         ("a_c + a%c(i)", "a_c + a%a_c_1(i)"),
+                                         ("b(b_c)%c(i)", "b(b_c)%b_c_1(i)"),
+                                         ("b(b_c)%c(i)", "b(b_c)%b_c_1(i)"),
+                                         ("a_b_c + a_b_c_1 + a%b%c",
+                                          "a_b_c + a_b_c_1 + a%a_b%a_b_c_2"),
+                                         ])
+def test_sym_writer_rename_members(fortran_reader, expressions):
+    '''Test that members are converted and get a unique name that
+    does not clash with any other variable used in the expression.
+
+    '''
+    # A dummy program to easily create the PSyIR for the
+    # expressions we need. We just take the RHS of the assignments
+    source = '''program test_prog
+                use my_mod
+                type(my_type) :: a, b(10)
+                integer :: i, j, x, a_c, b_c, a_b_c, a_b_c_1
+                x = {0}
+                end program test_prog '''.format(expressions[0])
+
+    psyir = fortran_reader.psyir_from_source(source)
+    expr = psyir.children[0].children[0].rhs
+    sympy_writer = SymPyWriter(list_of_expressions=[expr])
```

Comment:
This is what makes me uneasy about the fact we specify the same things to the constructor and then when calling the writer instance. I can understand the reason but it feels like we should be supplying the constructor with a list of reserved names or even a type map?

Reply:
@arporter Yes, I agree, that's not nice. It's hard to avoid - the SymPy writer needs to have all references (to avoid renaming something that will later when handling the second expression require to rename a symbol).
The only idea I have: I add a static function `convert_to_sympy` that takes a list of PSyIR nodes, and returns a list of strings containing each converted PSyIRnode. So this would mean we would do:
```
  exp1_str, exp2_str = SympyWriter.convert_to_sympy([exp1, exp2]).
```

The static method `convert_to_sympy` would then first collect all references and provide them to the SymPy constructor. I think it's 'nicer' to have this all in one class than in an additional utility function elsewhere (unless we need this functionality, but I don't think so atm).

Is that acceptable?

Reply:
I think it's better but it then changes the semantics with which the SympyWriter is called and thus it is no longer consistent with the other writers. How about:

    map = SympyWriter.create_type_map(expr)
    writer = SymPyWriter(map)
    lhs = writer(expr1)
    rhs = writer(expr2)
    ...
    str_exp1 = parse_expr(str_exp1, local_dict=map)

Since you have to separately supply a mapping to the `parse_expr` call anyway?

Reply:
Yes, but `map` is (potentially) modified when writing the two expressions (with the potential renaming of member names). While this would technically work (SympyWriter keeps a pointer to `map` and modifies it), but I think it's very unexpected that `map` is (potentially) modified after each `writer(exp1/2)` call :(

Plus, you are still supplying the arguments twice (argument of `create_type_map` must actually have both expressions to find all references :(  ).

What about:
```
map = SympyWriter.create_type_map(expr1, expr2)
writer = SymPyWriter()
lhs = writer(expr1, map)
rhs = writer(expr2, map)
```
This way there is at least some indication that map might be modified, though it still feels like a bad side effect (plus we break the normal interface of a writer). And we still supply each argument twice. 

Or:
```
writer = SympyWriter(exp1, exp2).
lhs = writer.get_expression(1)
rhs = writer.get_expression(2)
```
Not really good ... I'll leave this here in the spirit of wild brainstorming :)

I'll ponder about this tomorrow, but I am getting the feeling that we have to accept at least the fact that the arguments are specified twice :(

Reply:
I am going to implement your way, but also provide a convenience function that hides the detail from the users as a static method. This way I can well document the fact that the mapping is modified by the other calls to writer.
Won't be able to finish that today though :(

Reply:
I hadn't appreciated that the map would be modified by the calls to writer. Is there any way of avoiding this (by doing all the setup earlier)?

Reply:
Alternatively, would it help if we had a transformation that did something appropriate to the PSyIR before giving it to the writer? I've also just remembered that one of the problems I had when comparing index expressions was that my use of the Fortran writer got massively expensive because it copies the (whole?) tree each time. Is that a consideration here as well?

Reply:
> Is there any way of avoiding this (by doing all the setup earlier)?

Not really: we need to get all references (from all expressions) into the symbol table first (so that we don't rename any member into an existing variable that might only be used in the second expression). Then we need to handle all members. 
The only way to avoid this would be to do the whole conversion at the constructor (and then in the writer either do it again, or use cached results ... though caching results would also mean that we have to compare the trees to make sure we get the right cached result).

OK, now that I have written this: technically, after collecting all references, we could just do a `walk` in the constructor to find all members, add them to the symbol table and get new unique names for each member, and then store the new names for later. But that would just mean yet another walk across the whole tree.

> Alternatively, would it help if we had a transformation that did something appropriate to the PSyIR before giving it to the writer?

That's what I did originally (kind of ... it wasn't a stand-alone transformation) , and we all agreed that this way (copying the tree and changing the name of the members in the copied tree) is not the right way to g., and that it would be much easier to understand to use a writer.

I also think if we hide this from the user by providing a static function that takes a list of PSyIR expressions, and that returns a list of _sympified_ strings, that should be fine (if it is properly documented). This way the user does not even see that a SymPy writer is used.

We could provide a slightly different interface to the writer: instead of taking one expression, we let the writer convert a list (or tuple) of expressions all at the same time. Then we don't need this information at the constructor anymore, don't need to deal with keeping a state , ... - but it changes the interface for this writer (which I think is much worse than having a writer that keeps a state), and ... from a user's point of view it's pretty much the same as the aforementioned static function :)

TBH, I also don't see it as too much of a problem that there is one writer that keeps a state - that's why we have an OO approach. And I don't think a 'normal' user will ever use the SymPy writer: the SymbolicMaths class will provide all functions required for symbolic maths, and will hide the usage of the writer.

> I've also just remembered that one of the problems I had when comparing index expressions was that my use of the Fortran writer got massively expensive

Hmm - my initial thought was that for my use cases (dependency analysis) I will mostly be comparing indices, and while in theory they can be arbitrarily complex, in reality we will mostly see `i, j, k +- some constant`.  But if this had already been a bottleneck, this is a bit of a concern (I only remembered that converting expressions for error messages which were de-facto caught and ignored was expensive). Do you have any details (example code)? One thought I had: we might try to first throw the expression as is to sympy (without replacing members, kind constants with _) - if the most frequent use cases do not use structure, ... - that might be a bit cheaper. But I don't like that idea too much :(  One question would be: why do we actually copy the tree first? The writer is not supposed to modify anything.

I am happy to discuss this in the next telco, and on Monday I can just focus on creating more training material.

Reply:
It takes a copy (IIRC) because we have to flatten all of the symbol tables. The code where I hit problems was: https://github.com/stfc/PSyclone/blob/2fc39510a3bca5d07849dff5151e90600fc6b8aa/src/psyclone/psyir/nodes/array_mixin.py#L227-L252. Since SympyWriter is not attempting to produce Fortran it doesn't need to collapse the symbol tables and therefore we can probably avoid doing the tree copy. (Unless @sergisiso can remember that there are other reasons for doing the copy?).

Reply:
I also agree with you that we're probably best hiding everything in a static method - as you say, it's not really for public consumption.

Reply:
OK, I've implement your original suggestion:
```
map = SympyWriter.create_type_map(expr)
writer = SymPyWriter(map)
lhs = writer(expr1)
rhs = writer(expr2)
...
str_exp1 = parse_expr(str_exp1, local_dict=map)
```

But then also added as discussed a static convenience function. And while I originally let this function return strings (i.e. sympy-formatted representation of the PSyIR), I realised that I can take it a step further and let this convenience function return sympy expressions. The sympy-writer already imported sympy, so this change did not add a new dependency, and it also means that the `SymbolicMaths` class is (slightly) more portable. Also, if at some stage we should decide (e.g. for performance reasons) to avoid using a writer based on Fortran completely, we have one location where we could try to create the sympy tree based on PSyIR tree, without changing anything else (IIRC sympy has this functionality).